### PR TITLE
Case opt check

### PR DESCRIPTION
**Pull request by leonoverweel**
Added in `_compute_var_opt_mapping()` so that the check only happens once.

### CODE REVIEW

File: larq/optimizers.py [hunk 1]
```diff
@@ -210,6 +210,16 @@ def _compute_var_opt_mapping(self, grads_and_vars):
...
+        print("!~!!!!")
+        print(self.var_opt_mapping)
+
+        # Make sure that each optimizer touches at least one variable
+        for optimizer_index, (_, optimizer) in enumerate(self.pred_opt_pairs):
+            if optimizer_index not in self.var_opt_mapping.values():
+                raise ValueError(
+                    f"Optimizer `{optimizer}` did not claim any variables."
+                )
+
```
Comment (lgeiger) [inline larq/optimizers.py]: I think checking this in [`_apply_gradients()`](https://github.com/larq/larq/blob/c44ed976812edc2cc7d60c98155513942fc3af98/larq/optimizers.py#L136) would be simpler, then you could just do:
```python
for optimizer, opt_grads_and_vars in zip(self.optimizers, grad_var_lists):
    if not grad_var_lists:
        raise ValueError(f"Optimizer `{optimizer}` did not claim any variables.")
```

Comment (leonoverweel) [reply larq/optimizers.py]: Yes but then it happens on every training step; this way it just happens once when the var -> optimizer gets computed.

Comment (lgeiger) [reply larq/optimizers.py]: TensorFlow also raises errors in `_apply_gradients` too so it wouldn't be problematic, but this is fine too.

File: larq/optimizers_test.py [hunk 2]
```diff
@@ -82,7 +82,40 @@ def test_overlapping_predicates(self):
     def test_missing_default(self):
         with pytest.warns(Warning):
```
Comment (lgeiger) [inline larq/optimizers_test.py]: Shouldn't this be?
```suggestion
        with pytest.raises(ValueError):
```

Comment (leonoverweel) [reply larq/optimizers_test.py]: Nope, we [warn about this](https://github.com/larq/larq/blob/master/larq/optimizers.py#L209) instead of throwing.

Comment (leonoverweel) [reply larq/optimizers_test.py]: I think would be for scenario for example where you "freeze" FP weights by only passing an optimizer for binary weights.

File: larq/optimizers.py (hunk 1 above)
Comment (leonoverweel) [reply larq/optimizers.py]: I meant it mostly for efficiency.

File: larq/optimizers_test.py (hunk 2 above)
Comment (lgeiger) [reply larq/optimizers_test.py]: :+1: I think I misread the comment [here](https://github.com/larq/larq/pull/446/files#diff-4930c450271ff44e3368d2a51994982eR110)

Comment (lgeiger) [reply larq/optimizers_test.py]: Sorry the git diff just was weird to read.

**Pull request merged**